### PR TITLE
New task list version 1

### PR DESCRIPTION
**Pull request by louwie17**
Partly addresses woocommerce/woocommerce#32165 

Add some logic for easy switching of task list components, I need to convert the two column task list to Typescript for this. This logic should have with developing experiment 2 as long as we expect similar or the same props as `TaskListProps`.

Added an extra param called `options` to the task list that allows us to set custom options. I am setting a `task_past_tense` option that I make use of within each task to update the title when it is completed.
Also added a `tasklist-setup-experiment-1` feature flag (for now) until we integrate explat.

### Screenshots

Screen Shot 2022-03-10 at 11 55 21 AM

### Detailed test instructions:

- Enable the `tasklist-setup-experiment-1` feature flag in either the `config/development.json` and running the build process, or right in `includes/feature-config.php`
- Finish the onboarding flow and go to **WooCommerce > Home**
- The new task list should be present
- Completed tasks should be crossed through and updated in past tense (you can match each tasks wording with the Figma here: G9vi4e286yjdzyIdg8hBb1-fi-839%3A103364 )
- Step through each task and make sure things update accordingly.

### CODE REVIEW

**Comment by becdetat** [comment]
I haven't tested this, just looked through the code. There seem to be a couple of refactorings and changes to how things are done, I just want to check that the test instructions cover all of the code paths that have been touched? Avoiding regressions for example ;-)

File: client/tasks/tasks.tsx [hunk 1]
```diff
@@ -15,14 +15,37 @@ import { recordEvent } from '@woocommerce/tracks';
...
+const TaskList = lazy(
+	() => import( /* webpackChunkName: "task-list" */ './task-list' )
+);
+
+const TwoColumnTaskList = lazy(
+	() =>
+		import(
+			/* webpackChunkName: "two-column-task-list" */ '../two-column-tasks/task-list'
+		)
+);
```
Comment (octaedro) [inline client/tasks/tasks.tsx:36]: ❤️

File: src/Features/OnboardingTasks/Task.php [hunk 2]
```diff
@@ -116,6 +116,32 @@ public function get_parent_id() {
...
+			return array();
+		}
+		return $this->task_list->options;
+	}
+
+	/**
+	 * Get custom option.
+	 *
+	 * @param string $option_name name of custom option.
+	 * @return mixed
```
Comment (octaedro) [inline src/Features/OnboardingTasks/Task.php:135]: I think that we should put here: `@return mixed|null`

File: src/Features/OnboardingTasks/TaskLists.php [hunk 3]
```diff
@@ -91,8 +92,33 @@ public static function init_default_lists() {
...
+					'WooCommercePayments',
+					'Payments',
+					'Tax',
+					'Shipping',
+					'Marketing',
+					'Appearance',
+				),
+				'event_prefix' => 'tasklist_',
+				'options'      => array(
+					'task_past_tense' => true,
```
Comment (octaedro) [inline src/Features/OnboardingTasks/TaskLists.php:116]: I have the feeling that `task_past_tense` is not very intuitive, could we use something like `completed_title` or `use_completed_title` instead?

**Review by octaedro — APPROVED** [review]
You did a really nice job here @louwie17. I just left a couple of small comments, but the code looks good and it's testing well on my end, so I'll pre-approve this PR.

File: src/Features/OnboardingTasks/TaskLists.php (hunk 3 above)
Comment (louwie17) [reply src/Features/OnboardingTasks/TaskLists.php:116]: Thanks, I like `use_completed_title`

**Comment by louwie17** [comment]
@octaedro would you have time to look this over one more time? I made use of your suggestions, and added a couple extra tests.

File: client/two-column-tasks/task-list.tsx [hunk 4]
```diff
@@ -59,24 +87,17 @@ export const TaskList = ( {
...
-	const onDismissTask = ( { id, onDismiss } ) => {
-		dismissTask( id );
+	const onDismissTask = ( taskId: string, onDismiss?: () => void ) => {
+		dismissTask( taskId );
 		createNotice( 'success', __( 'Task dismissed' ), {
 			actions: [
 				{
 					label: __( 'Undo', 'woocommerce-admin' ),
-					onClick: () => undoDismissTask( id ),
+					onClick: () => undoDismissTask( taskId ),
```
[the remark this answers is not in the document]
Comment (octaedro) [reply client/two-column-tasks/task-list.tsx:100]: Gotcha! Scratch my comment 🙂

**Review by octaedro — APPROVED** [review]
Thank you for adding the tests @louwie17! Everything LGTM 🚀

**Pull request merged**